volmdlr/faces.py

```diff
@@ -5118,455 +5159,381 @@ def is_adjacent(self, face2: 'Face3D'):
+                new_faces_contours)
+        if isinstance(self, Triangle3D):
+            class_to_instanciate = PlaneFace3D
+        else:
+            class_to_instanciate = self.__class__
+        for contour, inner_contours in zip(new_faces_contours, new_inner_contours):
+            new_face = class_to_instanciate(self.surface3d, Surface2D(contour, inner_contours))
+            list_faces.append(new_face)
+        return list_faces
 
-        polygon2d = self.surface2d.outer_contour.to_polygon(angle_resolution=10)
-        border_distance, other_point = polygon2d.point_border_distance(point_2d, return_other_point=True)
+    def divide_face_with_closed_cutting_contours(self, list_closed_cutting_contours, list_faces):
+        """
+        Devides a Face3D with a list of Open cutting contours, that is, Contours going from one side
+        to another of the Face, or from the outer contour to one inner contour.
 
-        other_point = self.surface3d.point2d_to_3d(volmdlr.Point2D(*other_point))
+        :param list_closed_cutting_contours: list containing the closed cutting contours
+        :param list_faces: list of already divided faces
+        :return: list divided faces
+        """
+        for new_contour in list_closed_cutting_contours:
+            if len(new_contour.primitives) >= 3 and \
+                    new_contour.primitives[0].start == new_contour.primitives[-1].end:
+                inner_contours1 = [new_contour]
+                inner_contours2 = []
+                if list_faces:
+                    new_list_faces = self.get_closed_contour_divided_faces_inner_contours(list_faces, new_contour)
+                    list_faces = list_faces + new_list_faces
+                    continue
+                for inner_contour in self.surface2d.inner_contours:
+                    if new_contour.is_inside(inner_contour):
+                        inner_contours2.append(inner_contour)
+                        continue
+                    inner_contours1.append(inner_contour)
+                if isinstance(self, Triangle3D):
+                    class_to_instanciate = PlaneFace3D
+                else:
+                    class_to_instanciate = self.__class__
+                surf3d = self.surface3d
+                surf2d = Surface2D(self.surface2d.outer_contour, inner_contours1)
+                new_plane = class_to_instanciate(surf3d, surf2d)
+                list_faces.append(new_plane)
+                list_faces.append(class_to_instanciate(surf3d, Surface2D(new_contour, inner_contours2)))
+                continue
+            surf3d = self.surface3d
+            surf2d = Surface2D(self.surface2d.outer_contour, [])
+            if isinstance(self, Triangle3D):
+                class_to_instanciate = PlaneFace3D
+            else:
+                class_to_instanciate = self.__class__
+            new_plane = class_to_instanciate(surf3d, surf2d)
+            list_faces.append(new_plane)
+        return list_faces
```

Comment:
all this block has been displaced from PlaneFace3D to Face3D. Minor changes have been made, just to generalise the overall functionality.